### PR TITLE
Bitbucket Server Inline mode

### PR DESCRIPTION
**Pull request by codestergit**
### Todo:
- [x]  Proof of Concept
- [x]   Bitbucket Inline API implementation
- [x]  Convert BitBucketServerDiff to  GitStructuredDiff
- [x]  Platform changes
- [x]  Inline Template changes
- [x]  fix test cases
- [x]   Add test cases
- [x]   Add Changelog entry

### CODE REVIEW

**Comment by orta** [comment]
I'm pleasantly surprised at how little code this is. @sunshinejr I think you did some good architecting here to make this reasonably straight forwards.

@codestergit this PR looks 👍 to me, everything seems reasonable to me

**Comment by sunshinejr** [comment]
@orta yeah, I was expecting even less work but it all seems reasonable 👍 just few console logs here and there 😄

File: source/platforms/bitbucket_server/_tests/__snapshots__/_bitbucket_server_git.test.ts.snap [hunk 1]
```diff
@@ -28,6 +28,6 @@ exports[`the dangerfile gitDSL - BitBucket Server should include \`after\` text
 
 exports[`the dangerfile gitDSL - BitBucket Server should include \`before\` text content of the file 1`] = `"{\\"path\\":\\".gitignore\\",\\"repoSlug\\":\\"projects/PROJ/repos/repo\\",\\"ref\\":\\"d6725486c38d46a33e76f622cf24b9a388c8d13d\\"}"`;
 
-exports[`the dangerfile gitDSL - BitBucket Server should include \`removed\` text content of the file 1`] = `"node_modules*.log/test.py/.vscode.DS_Storecoverage.idea__pycache__/*.pyc*.swp"`;
+exports[`the dangerfile gitDSL - BitBucket Server should include \`removed\` text content of the file 1`] = `""`;
```
Comment (orta) [inline source/platforms/bitbucket_server/_tests/__snapshots__/_bitbucket_server_git.test.ts.snap]: Is this worth looking into?

Comment (codestergit) [reply source/platforms/bitbucket_server/_tests/__snapshots__/_bitbucket_server_git.test.ts.snap]: Yeah already checked it. 
Previously the logic was not correct. 
```
[{"destination":{"components":[".gitignore"],"extension":"gitignore","name":".gitignore","parent":"","toString":".gitignore"},"hunks":[{"destinationLine":1,"destinationSpan":11,"segments":[{"lines":[{"destination":1,"line":"node_modules","source":1,"truncated":false},{"destination":2,"line":"*.log","source":2,"truncated":false},{"destination":3,"line":"/test.py","source":3,"truncated":false},{"destination":4,"line":"/.vscode","source":4,"truncated":false},{"destination":5,"line":".DS_Store","source":5,"truncated":false},{"destination":6,"line":"coverage","source":6,"truncated":false},{"destination":7,"line":".idea","source":7,"truncated":false},{"destination":8,"line":"__pycache__/","source":8,"truncated":false},{"destination":9,"line":"*.pyc","source":9,"truncated":false}],"truncated":false,"type":"CONTEXT"},{"lines":[{"destination":10,"line":"","source":10,"truncated":false}],"truncated":false,"type":"ADDED"},{"lines":[{"destination":11,"line":"*.swp","source":10,"truncated":false}],"truncated":false,"type":"CONTEXT"}],"sourceLine":1,"sourceSpan":10,"truncated":false}],"source":{"components":[".gitignore"],"extension":"gitignore","name":".gitignore","parent":"","toString":".gitignore"},"truncated":false},{"destination":{"components":["banana"],"name":"banana","parent":"","toString":"banana"},"hunks":[{"destinationLine":1,"destinationSpan":1,"segments":[{"lines":[{"destination":1,"line":"bar","source":0,"truncated":false}],"truncated":false,"type":"ADDED"}],"sourceLine":0,"sourceSpan":0,"truncated":false}],"source":null,"truncated":false},{"destination":null,"hunks":[{"destinationLine":0,"destinationSpan":0,"segments":[{"lines":[{"destination":0,"line":"\"use strict\";","source":1,"truncated":false},{"destination":0,"line":"","source":2,"truncated":false},{"destination":0,"line":"module.exports = {","source":3,"truncated":false},{"destination":0,"line":"  runner: \"jest-runner-eslint\",","source":4,"truncated":false},{"destination":0,"line":"  displayName: \"lint\",","source":5,"truncated":false},{"destination":0,"line":"  testMatch: [\"<rootDir>/**/*.js\"],","source":6,"truncated":false},{"destination":0,"line":"  testPathIgnorePatterns: [\"node_modules/\"]","source":7,"truncated":false},{"destination":0,"line":"};","source":8,"truncated":false}],"truncated":false,"type":"REMOVED"}],"sourceLine":1,"sourceSpan":8,"truncated":false}],"source":{"components":["jest.eslint.config.js"],"extension":"js","name":"jest.eslint.config.js","parent":"","toString":"jest.eslint.config.js"},"truncated":false}]
```

This is diff as bitbucket json . If you see there is no `removed`  hunks for `.gitignore` file.

**Comment by codestergit** [comment]
@orta 

Found https://github.com/danger/danger-js/issues/563.

Opened new pr for fix https://github.com/danger/danger-js/pull/564. If you want I can add fix here.

**Comment by codestergit** [comment]
@orta 
Everything working good. It is good to go.

Please review. Please find the results.

warn

fail

File: source/platforms/bitbucket_server/BitBucketServerAPI.ts [hunk 2]
```diff
@@ -209,7 +262,7 @@ export class BitBucketServerAPI {
   }
 
   get = (path: string, headers: any = {}, suppressErrors?: boolean): Promise<node_fetch.Response> =>
-    this.api(path, headers, undefined, "GET", suppressErrors)
+    this.api(path, headers, null, "GET", suppressErrors)
```
Comment (orta) [inline source/platforms/bitbucket_server/BitBucketServerAPI.ts]: ah nice! Yeah

File: source/platforms/bitbucket_server/BitBucketServerGit.ts [hunk 3]
```diff
@@ -98,14 +97,22 @@ const bitBucketServerDiffToGitJSONDSL = (diffs: BitBucketServerDiff[], commits:
...
+      changes: hunk.segments
+        .map(segment =>
+          segment.lines.map(line => ({
+            type: segmentValues[segment.type] as "add" | "del" | "normal",
+            content: line.line,
+            sourceLine: line.source,
+            destinationLine: line.destination,
+          }))
+        )
+        .reduce((a, b) => a.concat(b), []),
```
Comment (orta) [inline source/platforms/bitbucket_server/BitBucketServerGit.ts]: good work here ^

File: source/platforms/bitbucket_server/_tests/_bitbucket_server_api.test.ts [hunk 4]
```diff
@@ -161,19 +161,47 @@ describe("API testing - BitBucket Server", () => {
 
     expect(api.fetch).toHaveBeenCalledWith(
       `${host}/rest/api/1.0/projects/FOO/repos/BAR/pull-requests/1/activities?fromType=COMMENT`,
-      { method: "GET", body: {}, headers: expectedJSONHeaders },
+      { method: "GET", body: null, headers: expectedJSONHeaders },
```
Comment (orta) [inline source/platforms/bitbucket_server/_tests/_bitbucket_server_api.test.ts]: thanks, I missed these!

File: CHANGELOG.md [hunk 5]
```diff
@@ -13,6 +13,10 @@
 
 ## Master
 
+## 3.4.8
```
Comment (orta) [inline CHANGELOG.md]: please avoid version changes, I do that as I ship a new one and you'll just get unnecessary merge conflicts

File: source/runner/templates/_tests/_bitbucketServerTemplate.test.ts [hunk 6]
```diff
@@ -43,28 +43,28 @@ describe("generating messages for BitBucket server", () => {
...
 
   it("Shows the message", () => {
-    const issues = inlineTemplate(messagesResults)
-    expect(issues).toContain("- :book: Message")
-    expect(issues).not.toContain("- :no_entry_sign:")
-    expect(issues).not.toContain("- :warning:")
+    const issues = inlineTemplate("blankID", messagesResults, "File.swift", 5)
+    expect(issues).toContain("- 📖 Message")
+    expect(issues).not.toContain("- 🚫")
+    expect(issues).not.toContain("- ⚠️")
```
Comment (orta) [inline source/runner/templates/_tests/_bitbucketServerTemplate.test.ts]: cool

**Comment by orta** [comment]
Ah, oops, I just made a release thinking I merged this PR 👯‍♂️

**Comment by orta** [comment]
OK, will ship this as `2.5.1` - hah

**Comment by codestergit** [comment]
I think  you mean `3.5.1`

**Comment by sunshinejr** [comment]
Oh, I've missed the merge! Great job @codestergit, thanks for adding this! 🎉

**Pull request merged**